Add 'auto' tactic and STLC example

pwang347 opened the pull request:
## Description
This change introduces a Coq-like `auto` tactic using user-defined hints and the existing theorem context, as well as a test case for formalized STLC.

At the moment the `auto` tactic does a simple BFS brute force scan with minor pruning, but it may be possible to introduce some more intelligent pattern matching or tactic/id re-ordering in the future.

## Points to note/for discussion:
* I benchmarked the `auto.rkt` test file to take around 30 seconds for proving five (trivial) theorems from scratch, with some taking a bit longer than others (inductive example seems to work almost instantaneously, for instance)
* As it is, the hints need to be cleared each time after use otherwise they bleed into the subsequent auto runs
* I'm not entirely sure if the changes related to apply and by-intros was necessary, though I'm not quite sure how to work with syntax methods that aren't tactics operating on a tree zipper.

## Tests performed:
* Just ran the two test files and saw no errors

pwang347 committed:
Add auto tactic

pwang347 committed:
Add auto tactic and STLC example

pwang347 committed:
Fix requires

pwang347 committed:
Minor formatting

pwang347 committed:
Minor

pwang347 commented on cur-lib/cur/ntac/hint.rkt:
```diff
@@ -0,0 +1,126 @@
+
+  ; bypass phase-level restriction in hints-add!
+  (define (set-hints! v)
+    (set! hints (make-parameter v)))
```
`(hints (cons v (hints))`

pwang347 commented on cur-lib/cur/ntac/hint.rkt:
```diff
@@ -0,0 +1,126 @@
+    tz)
+
+  (define (clear-hints! tz)
+    (set-hints! null)
```
`(hints '())`

pwang347 commented on cur-lib/cur/ntac/hint.rkt:
```diff
@@ -0,0 +1,126 @@
+  (define-syntax (hints-add! syn)
+    (syntax-case syn ()
+      [(_ H) #'(begin
+                 (set-hints! (cons #'H (hints)))
```
`(set-hints! #'H)`

pwang347 commented on cur-lib/cur/ntac/hint.rkt:
```diff
@@ -0,0 +1,126 @@
+                       #:break (not (false? solution))
+                       (filter (compose not false? car)
+                               (for/list ([thm-name thm-names])
+                                 #:break (not (false? solution))
```
`#:break solution`

stchang commented on cur-test/cur/tests/stlc-formal.rkt:
```diff
@@ -0,0 +1,344 @@
+  (by-apply v-abs)
+  simpl
+  (by-apply multi-refl))
+
```
So none of the STLC examples are working? Did you try to give `by-apply` the explicit arguments from the explicit proof term below?

wilbowma replied:
I asked Paul to put the STLC test in there, even though most of the tactic proofs don't work yet. We could make it work manually, but this formalism was a direct port from a Coq tutorial and the goal was to explore limitations in Cur more than to actually make it work. Even so, the largeish formalism and things that do work are good tests, and are slow, so a good benchmark.

stchang replied:
Ok cool. I'll try to check it out later to see if there's an obvious fix.

wilbowma replied:
The problem is that we don't implement `eapply` and the Coq proofs used that to automagically figure out the arguments. We could probably fix is pretty quickly as you suggested.

wilbowma commented:
We're going to make the hint database automatically lexically scoped by redefining `ntac` and `define-theorem` to parameterize the hint database in the scope of a new ntac call, and fix a few code smells before merging.

pwang347 committed:
Add subgoal completion, some other misc fixes

pwang347 committed:
Add newline

pwang347 committed:
add top-level depth and fn parameter

wilbowma committed:
Added STLC model

wilbowma committed:
Style fixes

- Remove debugging code
- Change name of parameter to current-$name
- Use parameterize more
- Deprecate imperative interfaces in favor of parameters interface

wilbowma commented:
Pushed some changes to this in my branch `auto`; take a look.

pwang347 committed:
Merge branch 'auto' of github.com:wilbowma/cur into pwang-auto-stlc

pwang347 committed:
Minor fixes

pwang347 commented:
> Pushed some changes to this in my branch `auto`; take a look.

Merged in the changes. Just to be clear, it's fine that we have `begin-for-syntax` blocks for the parameters right?

For example:
```
(begin-for-syntax
  (current-max-auto-depth 4))
```

And the remaining TODO comments to remove the methods are so that we only let the user set hints at the top level similar to the max auto depth?

pwang347 committed:
Skip test

wilbowma commented:
Something like this interface, where scope manages removing hints automagically.
```racket
(define-theorem ...
  #:hints (e ...)

  (hints e1 ...
    auto
    simpl))
```

pwang347 committed:
Update hinting mechanism

pwang347 committed:
Should probably still keep the nttz-done

wilbowma commented:
I've rebased, squashed commits, and fixed some stylistic things more in origin/auto.

We should add documentation for this. Most of the documentation is out of date, but the ntac stuff is still correct, so it' be good to stay on top of it. Think you can figure out how to do that?

pwang347 commented:
> I've rebased, squashed commits, and fixed some stylistic things more in origin/auto.
> 
> We should add documentation for this. Most of the documentation is out of date, but the ntac stuff is still correct, so it' be good to stay on top of it. Think you can figure out how to do that?

Will do - I'll merge in your changes and add documentation. Thanks!

wilbowma commented:
You may not want to merge my changes since I’ve done a lot of rebasing and squashing; I can always merge documentation and such.

-- 
Sent from my phoneamajig

> On Oct 18, 2019, at 10:48, Paul <notifications@github.com> wrote:
> 
> I've rebased, squashed commits, and fixed some stylistic things more in origin/auto.
> 
> We should add documentation for this. Most of the documentation is out of date, but the ntac stuff is still correct, so it' be good to stay on top of it. Think you can figure out how to do that?
> 
> Will do - I'll merge in your changes and add documentation here. Thanks!
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or unsubscribe.

pwang347 committed:
Add documentation

pwang347 commented:
Current state of affairs for the documentation:
* Some red squiggles for `forall`, `->`, `Nat`, `define`, I'm guessing it'll stay this way until we add documentation for these again?
* Some weird documentation behaviour with duplicate names across base and hints. For now, I've made define-theorem point to the one defined in hints while everything else (e.g. ntac) points to base as before.
* Removed some of the unused tactics from the documentation (I made sure to check commit history for the files, and it looks like these variants were indeed removed)
* by-obvious wasn't working on types since they aren't considered IDs
* Bumped cur-lib version number
* I realized that whatever importing hints probably doesn't need to bother with base unless it's specifically using the syntax-phase constructs

Please let me know if I should add anything else!

wilbowma commented:
Looks like the blocking issue in Racket was fixed via https://github.com/racket/racket/issues/2981

wilbowma commented:
Merged in e86c676925641bd4ece88ba3972ae0c80561cb68
